### PR TITLE
Ksz8851snl driver for Ethernet-XPRO

### PR DESCRIPTION
**Pull request by poppe34**
This is just the driver for the ksz8851snl... I still need to create the board folder. it also has changes to the default example to aid with development.  That example should be removed after driver validation.

### CODE REVIEW

File: drivers/include/ksz8851snl.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
```
Comment (OlegHahm) [inline drivers/include/ksz8851snl.h]: Are you intensionally granting Kaspar the Copyright here?

File: drivers/ksz8851snl/include/ksz8851snl_config.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (C) 2015
```
Comment (OlegHahm) [inline drivers/ksz8851snl/include/ksz8851snl_config.h]: Copyright holder is missing.

File: drivers/ksz8851snl/include/ksz8851snl_regs.h [hunk 3]
```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (C) 2015
```
Comment (OlegHahm) [inline drivers/ksz8851snl/include/ksz8851snl_regs.h]: Copyright holder is missing.

**Comment by OlegHahm** [comment]
Cool! I'm eager to test. Two comments:
1.) Please rebase to master to get rid of the merge commits (if you need assistance with Git, I'm happy to help).
2.) Your application would fit better in hour `tests` folder (as a kind of "smoke test").

**Comment by poppe34** [comment]
I left the copyright open because I wasn't sure if I just put my name or what exactly I do.  I will add my name and email to the files.  As far as the git I will look into the rebasing.  I will ask if/when I run into questions.

**Comment by OlegHahm** [comment]
The copyright line is mostly needed for US laws. For most European laws you as the author holds the copyright anyway.

**Comment by poppe34** [comment]
Sorry I am trying to figure out the rebasing

**Comment by poppe34** [comment]
Ok I am lost what should I be doing.  I rebase it to get rid of the merge comments.  But that doesn't seem to do it. Then it uploaded a bunch of other comments too.  What process should I follow to properly rebase.

**Comment by OlegHahm** [comment]
If you use your favorite search engine for the terms `git rebase tutorial` you should find some helpful hints.
What I find always very helpful is to call `git rebase -i origin/master` from the corresponding branch. This should open an editor with all the commits that are not part of master. Now you reorder, remove, edit or squash these commits. Once you close the editor, git will perform the rebasing. If there is a merge-conflict, it will tell you and you can run `git merge-tool` to resolve the conflicts. In the end you will have to force push `git push -f <remote> <branch>` your changes. If you need live help, I'm also available at IRC tonight.

**Comment by poppe34** [comment]
Ok I rebased but I am sure it is totally wrong.  If i need to create a new branch make the changes then re-pull I can.  It won't let me do any more rebasing so I am just hoping this is close to right.

**Comment by BytesGalore** [comment]
@poppe34 seems so, it looks like you merged commits from others to your branch. 
But don't worry, this can be fixed :).

I guess the easiest way for you is to: 
1. move your ~7 commits to a new clean branch [1]
2. revert all changes in this branch [2], for your situation you would reset back to [3] and rebase this "roll-backed" branch
3. finally move your commits back to this (basically the same as 1.)

I hope this helps a bit

[1] http://stackoverflow.com/questions/1628563/move-the-most-recent-commits-to-a-new-branch-with-git/1628584#1628584
[2] http://stackoverflow.com/questions/1616957/how-do-you-roll-back-reset-a-git-repository-to-a-particular-commit/1616959#1616959
[3] https://github.com/RIOT-OS/RIOT/commit/e72bde5805abdc59f6342fee52dd4e3336900772

**Comment by poppe34** [comment]
Thanks for all the help.  I have started a new branch and I will test it out and make sure all files were updated.  I will submit the pull then.

**Comment by msolters** [comment]
Hey @poppe34, awesome work!

I'm sure you've made some changes to it by now, but FWIW I pulled your changes into a "cleaner" [branch here](https://github.com/msolters/RIOT/tree/ksz8851snl) if it helps any.

That's what I used to build and test your driver, anyhow, and again -- awesome work!  Doesn't seem to like 8Mhz though -- it crashes immediately after "Link Up" if I disable the PLL clock, when it goes looking for nonexistant NETOPT 26.

**Comment by poppe34** [comment]
I worked more on this weekend, and I am running into some issues.  I don't know if any changes to the core that effected my original driver, but I am testing.  It seems to be a mutex issue, somewhere it is locking and not releasing.  I will squash that bug and submit the pull.  Thanx for the the kind words msolters.  I don't see why it wouldn't work at 8Mhz other than I have the driver running at 5Mhz over spi.  Did you lower the baud when you lowered the mcu core clock?

**Comment by msolters** [comment]
@poppe34 I just tried again at 8Mhz.  Here's the behaviour.

SAM boots up, loads an Iface corresponding to the ethernet adapter.  Gets an IPV6 address, everything is fine.

Every few seconds I check the output of `ifconfig` to see what's going on.  Things are fine for a minute or more.

Inevitably, though, one of the `ifconfig` calls shows that, suddenly, the ethernet Iface disappears!  At that point, the SAM will stop responding to *anything* over EDBG serial.  Restarting the SAM restores serial, but then the problem only repeats itself.

I'd like to point out that EDBG serial only stops working when I call `ifconfig`, and the Iface is broken.  In other words, even though the Iface is presumably down, I could continue to call `help` or `ps` as much as I want.  But as soon as `ifconfig` gets called, EDBG serial breaks!

**Comment by msolters** [comment]
Here's a dump of what the ethernet Iface reports right before it goes missing:

```
2015-11-02 13:54:07,000 - INFO # Iface  6   HWaddr: 00:04:25:19:1a:3e [KSZ8851 Driver]:Device Get Opt Called and not handled: 26
2015-11-02 13:54:07,001 - INFO #  State: IDLE
2015-11-02 13:54:07,003 - INFO #            MTU:1280  
2015-11-02 13:54:07,006 - INFO #            Source address length: 6
2015-11-02 13:54:07,008 - INFO #            Link type: wired
2015-11-02 13:54:07,014 - INFO #            inet6 addr: ff02::1/128  scope: local [multicast]
2015-11-02 13:54:07,019 - INFO #            inet6 addr: fe80::204:25ff:fe19:1a3e/64  scope: local
2015-11-02 13:54:07,026 - INFO #            inet6 addr: ff02::1:ff19:1a3e/128  scope: local [multicast]
2015-11-02 13:54:07,031 - INFO #            inet6 addr: ff02::2/128  scope: local [multicast]
```

Is `[KSZ8851 Driver]:Device Get Opt Called and not handled: 26` something you were seeing wrt the mutex issue as well?  I'd hope they are one and the same problem.

**Comment by poppe34** [comment]
I am too seeing that issue as well, but I am seeing it at all clock speeds now.  Something must have changed in the core since I originally wrote this and it is conflicting.  I will try to squash it.

please close this branch and refer to pull request:  
initial ksz8851snl driver files for Ethernet XPRO expansion #4212

**Comment by msolters** [comment]
:+1:

**Comment by haukepetersen** [comment]
Closed (on request) in favor of #4212

**Pull request closed without merging**